### PR TITLE
Read serialised TB data

### PR DESCRIPTION
**Pull request by PhilipVinc**
I'd love some feedback on this one @oxinabox @tkf 

This one allows you to do 
```Julia
using TensorBoardLogger, Logging, ValueHistories

tb_logger = TBLogger("testlog", tb_overwrite)

with_logger(tb_logger) do
    for i=1:5
        @info "test" val=i b=I*2 mat=rand(3)
    end
end

hist = MVHistory()

TensorBoardLogger.map_summaries(tb_logger) do tag, iter, val
    push!(hist, Symbol(tag), iter, val)
end

julia> hist
MVHistory{History}
  :test/val => 5 elements {Int64,Float32}
  :test/b => 5 elements {Int64,Float32}
```

Main issue are arrays: currently we store as `Histograms` both histograms of data and flattened N-d matrices. When we read those back, we can't really tell what was what before serialization.
I think I could add some extra data to the ProtoBuffer and store the shape of the original matrix, in the unused `metadata` field. This would be compatible with TensorBoard, would allow us to correctly de-serialize data, but it's not something in the standard protocol...

Any thoughts?

The implementation is a bit convoluted to support multiple files and purge/restart.
It would also be nice to also implement an iterator to do
```julia
for (tag, item, val) = tb_logger
  #dostuff
end
``` 
but I did not find non-ugly ways to do this...

### CODE REVIEW

File: src/deserialization.jl [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+"""
+    TBEventFileCollectionIterator(path; purge=true)
+
+Iterate along all event-files in the folder `path`.
+When the keyword argument `purge==true`, if the i+1-th file begins with a purge
+at step `s`, the i-th file is read only up to step `s`.
+"""
+struct TBEventFileCollectionIterator
+    dir::String
+    files::Vector
```
Comment (oxinabox) [inline src/deserialization.jl]:
```suggestion
    files::Vector{String}
```

**Comment by oxinabox** [comment]
Doesn't look too crazy to me, but I only had a quick glance

File: src/logger_dispatch.jl [hunk 2]
```diff
@@ -75,15 +75,14 @@ summary_impl(name, value::Any) = text_summary(name, value)
 
 
 ########## For things going to LogHistograms ########################
-preprocess(name,   val::AbstractVector{<:Real}, data) = push!(data, name=>val)
-summary_impl(name, val::AbstractVector{<:Real}) = histogram_summary(name, collect(0:length(val)),val)
-
 preprocess(name,   (bins,weights)::Tuple{AbstractVector,AbstractVector}, data) = push!(data, name=>(bins, weights))
 summary_impl(name, (bins,weights)::Tuple{AbstractVector,AbstractVector}) = histogram_summary(name, bins, weights)
 
+preprocess(name, val::AbstractArray{<:Real}, data) = push!(data, name=>val)
```
Comment (shashikdm) [inline src/logger_dispatch.jl:81]: This is causing build test to fail. I think the test should change from:
```
@test last(data[4]) == vec(vals)
```
to
```
@test last(data[4]) == vals
```

**Comment by PhilipVinc** [comment]
Current issues:
 - De-serialization is sequential and works on one SummaryValue at a time
 - Array{Colorant,3} (so 3-d images like MRI data) are serialised as a sequence of 2D images with the same 'step' so that tensor board automatically combines them. But so there are N=size(image, 3) distinct summaries.

So those are read back as N different images with the same step. I should find a way to recombine them smartly. Maybe when we log images, we look ahead at the next summary?

Text:
 - Right now we use `markdown_repr` so that all strings like `foo="astring"` is logged as `markdown_repr(foo) ==> "\"astring\"" `. We should get rid of the leading `\"`.

**Comment by PhilipVinc** [comment]
If we fix the thing of 3-dimensional images, we might as well do the same for complex-valued numerical arrays, and recombine /re and /im parts into a single one.

**Comment by PhilipVinc** [comment]
Mac tests work... it's really just a linux QT problem, triggered by FileIO load.
It's weird. Because we have no UI. I'm simply using FileIO to deserialise a binary blob...

**Comment by PhilipVinc** [comment]
It's because of [FileIO#174](https://github.com/JuliaIO/FileIO.jl/issues/174)

**Comment by PhilipVinc** [comment]
YAY!! Build successful!

**Comment by jessebett** [comment]
Yay!

**Pull request merged**